payload/reggae/rules/cmake.d

```diff
@@ -0,0 +1,320 @@
+        }
+    });
+
+    return [flags, libraries, libraryPath].join(" ");
+}
+
+private enum TargetType : string {
+    Executable = "EXECUTABLE",
+    StaticLib = "STATIC_LIBRARY",
+    SharedLib = "SHARED_LIBRARY",
+}
+
+private bool isSupportedTarget(in string target) {
+    import std.traits : EnumMembers;
+    import std.algorithm : canFind;
+    return [EnumMembers!TargetType].canFind(target);
+}
+
+private imported!"reggae.build".Target toReggaeTarget(in imported!"std.json".JSONValue target,
+                                                      CMakeInfo cmakeInfo) {
+    import reggae.config: options;
+    import reggae.rules.common : objectFile, link, staticLibraryTarget;
+    import reggae.build : Target;
+    import reggae.types : SourceFile, ExeName, Flags;
+    import std.format : format;
+    import std.exception : enforce;
+    import std.path : buildNormalizedPath, isAbsolute, baseName;
+    import std.range : empty;
+    import std.stdio : writeln, stderr;
+
+    const sourceDir = cmakeInfo.codeModel["paths"]["source"].str;
+    const buildDir = cmakeInfo.codeModel["paths"]["source"].str;
+
+    Target[] intermediateTargets;
+
+    foreach (sourceJsonObj; target["sources"].array) {
+        const filePath = sourceJsonObj["path"].str;
+        const sourceFilePath = filePath.isAbsolute ? filePath : buildNormalizedPath(sourceDir, filePath);
+
+        const needsToBeCompiled = "compileGroupIndex" in sourceJsonObj;
+        if (!needsToBeCompiled) {
+            enforce(sourceFilePath.isHeaderFile,
+                    "File '%s' which is neither compiled nor a header is not supported.".format(sourceFilePath));
+            continue;
+        }
+
+        const compileGroupIndex = sourceJsonObj["compileGroupIndex"].integer;
+        auto compileOptions = target.getCompileOptions(compileGroupIndex);
+
+        intermediateTargets ~= objectFile(options, // options - why is it needed / is it the right thing to do?
+                                          SourceFile(sourceFilePath),
+                                          compileOptions.compilerFlags,
+                                          compileOptions.includes);
+    }
+
+    enforce(target["type"].str.isSupportedTarget,
+            "Target type '%s' not supported.".format(target["type"].str));
+
+    auto artifacts = target["artifacts"].array;
+    enforce(artifacts.length >= 1, "Supported targets must generate at least one artifact.");
+    if (artifacts.length > 1) {
+        stderr.writeln("[Warning] Target '" ~ target["name"].str ~ "' generates multiple artifacts.");
+    }
+
+    string artifactFilePath;
+    foreach (art; artifacts) {
+        if (art["path"].str.baseName == target["nameOnDisk"].str) {
+            artifactFilePath = art["path"].str;
+            break;
+        }
+    }
+    enforce(!artifactFilePath.empty, "Could not find the expected artifact path.");
+
+    // TODO is this the right artifact path?
```

Comment:
The question here is whether we should fully replicate the paths used by `CMake` (i.e. even for the artifacts) or if in some cases `reggae` might expect them to be in some specific place. I guess the latter scenario is pretty unlikely, but thought it would be good to ask.

Reply:
My first instinct is to say "probably not".